ogr/ogrsf_frmts/xodr/ogrxodrdatasource.cpp

```diff
@@ -0,0 +1,186 @@
+ ****************************************************************************/
+
+#include "ogr_xodr.h"
+using namespace odr;
+using namespace pugi;
+using namespace std;
+
+CPL_CVSID("$Id$")
+
+OGRXODRDataSource::OGRXODRDataSource()
+{
+    layers = nullptr;
+    nLayers = 0;
+}
+
+OGRXODRDataSource::~OGRXODRDataSource()
+{
+    for (int i = 0; i < nLayers; i++)
+    {
+        delete layers[i];
+        layers[i] = nullptr;
+    }
+    CPLFree(layers);
+}
+
+int OGRXODRDataSource::Open(const char *fileName, char **openOptions,
+                            int bUpdate)
+{
+    VSILFILE *file = nullptr;
+
+    bool updatable = CPL_TO_BOOL(bUpdate);
+    if (updatable)
+    {
+        CPLError(CE_Failure, CPLE_OpenFailed,
+                 "Write or update access not supported by XODR driver.");
+        return CE_Failure;
+    }
+    else
+    {
+        file = VSIFOpenExL(fileName, "r", true);
+    }
+
+    if (file == nullptr)
+    {
+        //TODO is this ever called on an opening error? An incorrect file name or path is caught earlier already.
+        CPLError(CE_Failure, CPLE_OpenFailed,
+                 "Failed to load OpenDRIVE file %s.", fileName);
+        return CE_Failure;
+    }
+
+    const char *openOptionValue =
+        CSLFetchNameValueDef(openOptions, "EPS", "1.0");
+    eps = CPLAtof(openOptionValue);
+    openOptionValue = CSLFetchNameValueDef(openOptions, "DISSOLVE_TIN", "NO");
+    dissolveTIN = CPLTestBool(openOptionValue);
+
+    odr::OpenDriveMap xodr(fileName, false);
+    std::string proj4Defn = xodr.proj4;
+    std::vector<odr::Road> roads = xodr.get_roads();
+    RoadElements roadElements = createRoadElements(roads);
+
+    nLayers = 6;
+    //TODO Do we have to update this, taking into account all different layer subclasses?
+    layers =
+        (OGRXODRLayer **)CPLRealloc(layers, sizeof(OGRXODRLayer *) * nLayers);
+    layers[0] = new OGRXODRLayerReferenceLine(roadElements, proj4Defn);
+    layers[1] = new OGRXODRLayerLaneBorder(roadElements, proj4Defn);
+    layers[2] = new OGRXODRLayerRoadMark(roadElements, proj4Defn, dissolveTIN);
+    layers[3] = new OGRXODRLayerRoadObject(roadElements, proj4Defn);
+    layers[4] = new OGRXODRLayerLane(roadElements, proj4Defn, dissolveTIN);
+    layers[5] =
+        new OGRXODRLayerRoadSignal(roadElements, proj4Defn, dissolveTIN);
+    return TRUE;
+}
+
+OGRLayer *OGRXODRDataSource::GetLayer(int iLayer)
+{
+    if (iLayer < 0 || iLayer >= nLayers)
+        return NULL;
+
+    return layers[iLayer];
+}
+
+int OGRXODRDataSource::TestCapability(CPL_UNUSED const char *capability)
+{
+    if (EQUAL(capability, ODsCCreateLayer))
+        return FALSE;
+    else if (EQUAL(capability, ODsCDeleteLayer))
+        return FALSE;
+    else if (EQUAL(capability, ODsCZGeometries))
+        return TRUE;
+    return FALSE;
+}
+
+RoadElements
+OGRXODRDataSource::createRoadElements(const std::vector<odr::Road> roads)
+{
+    RoadElements elements;
+
+    for (odr::Road road : roads)
```

Comment:
const ref if possible:
```suggestion
    for (const odr::Road &road : roads)
```
applies to inner loops as well